internal/access/credential.go

```diff
@@ -12,156 +12,152 @@ import (
-		// provider_user if it's missing.
-		_, _ = data.CreateProviderUser(rCtx.DBTxn, data.InfraProvider(rCtx.DBTxn), user)
+	if err := updateCredential(tx, credential, newPassword); err != nil {
+		return "", err
 	}
 
-	return nil
+	return newPassword, nil
 }
 
-func updateCredential(c *gin.Context, user *models.Identity, newPassword string, isSelf bool) error {
+// UpdateCredential updates a user's password to a specified valued. In order to update the user's password,
+// specified requirements must be met:
+//
+// 1. The old password hash must match value stored in the database
+// 2. The new password must meet the password policy defined for the organization
+func UpdateCredential(c *gin.Context, user *models.Identity, oldPassword, newPassword string) error {
 	rCtx := GetRequestContext(c)
-	db := rCtx.DBTxn
+	tx := GetRequestContext(c).DBTxn
 
-	if err := checkPasswordRequirements(db, newPassword); err != nil {
-		return err
-	}
+	errs := make(validate.Error)
 
-	if err := checkBadPasswords(newPassword); err != nil {
-		return err
+	credential, err := data.GetCredentialByUserID(tx, user.ID)
+	if err != nil {
+		return fmt.Errorf("get credential: %w", err)
 	}
 
-	hash, err := bcrypt.GenerateFromPassword([]byte(newPassword), bcrypt.DefaultCost)
-	if err != nil {
-		return fmt.Errorf("hash: %w", err)
+	// compare the stored hash of the user's password and the hash of the presented password
+	if err := bcrypt.CompareHashAndPassword(credential.PasswordHash, []byte(oldPassword)); err != nil {
+		errs["oldPassword"] = append(errs["oldPassword"], "invalid password")
+		return errs
 	}
 
-	userCredential, err := data.GetCredentialByUserID(db, user.ID)
-	if err != nil {
-		if errors.Is(err, internal.ErrNotFound) && !isSelf {
-			if err := data.CreateCredential(db, &models.Credential{
-				IdentityID:      user.ID,
-				PasswordHash:    hash,
-				OneTimePassword: true,
-			}); err != nil {
-				return fmt.Errorf("creating credentials: %w", err)
-			}
-			return nil
-		}
-		return fmt.Errorf("existing credential: %w", err)
+	if err := checkPasswordRequirements(tx, newPassword); err != nil {
+		return err
 	}
 
-	userCredential.PasswordHash = hash
-	userCredential.OneTimePassword = !isSelf
+	credential.OneTimePassword = false
 
-	if err := data.UpdateCredential(db, userCredential); err != nil {
-		return fmt.Errorf("saving credentials: %w", err)
+	if err := updateCredential(tx, credential, newPassword); err != nil {
+		return err
 	}
 
-	if isSelf {
-		// if we updated our own password, remove the password-reset scope from our access key.
-		if accessKey := rCtx.Authenticated.AccessKey; accessKey != nil {
-			accessKey.Scopes = sliceWithoutElement(accessKey.Scopes, models.ScopePasswordReset)
-			if err = data.UpdateAccessKey(db, accessKey); err != nil {
-				return fmt.Errorf("updating access key: %w", err)
+	if accessKey := rCtx.Authenticated.AccessKey; accessKey != nil {
```

Comment:
```suggestion
        // if we updated our own password, remove the password-reset scope from our access key.
        if accessKey := rCtx.Authenticated.AccessKey; accessKey != nil {
```

This is confusing without the comment imo